Change CC0001 so that it does not apply to primitives

Change `CC0001` so that it only works on non primitives. Raise the new `CC0105` for primitives.
There is a list of primitives types [on MSDN](https://msdn.microsoft.com/en-us/library/system.type.isprimitive%28v=vs.110%29.aspx) of primitives from C#:

> The primitive types are Boolean, Byte, SByte, Int16, UInt16, Int32, UInt32, Int64, UInt64, IntPtr, UIntPtr, Char, Double, and Single.

**Remarks**:
* DateTime is not a primitive and should still be covered by `CC0001`.
* String is not a primitive and should still be covered by `CC0001`.
* `CC0105` Should apply to both when the type is declared using an identifier name (`System.Int32 i;`) or using a predefined type (`int i;`).
* `CC0105` should follow `CC0001` category and severity level (Style and Warning).
* The code fix provider for `CC0001` should updated to support `CC0105`, but not other change should be made to it.

Just as a reference, this is @OsirisTerje original comment:

> As it is now, the CC0001 rule, "Use 'var' instead of specifying the type name" applies to any type.  It could be wise to separate this into two rules, or, if possible, have an option for it, so that one can have it apply to only non-primitive types. Resharper also have this as an option.

I agree that it would be wise to separate this into two rules. This var usage rule applied to number types can induce to errors. For example, the following sentence will launch the compiler error CS0031:

> short c = 1000000000; //CS0031

If you use the following sentence nothing will be launched:

> var c = 1000000000;

--- a/src/CSharp/CodeCracker/Style/AlwaysUseVarAnalyzer.cs
+++ b/src/CSharp/CodeCracker/Style/AlwaysUseVarAnalyzer.cs
@@ -16,7 +16,7 @@ namespace CodeCracker.CSharp.Style
         const string Description = "Usage of an implicit type improve readability of the code.\r\n"
             + "Code depending on types for their readability should be refactored with better variable "
             + "names or by introducing well-named methods.";
-        internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+        internal static readonly DiagnosticDescriptor RuleNonPrimitives = new DiagnosticDescriptor(
             DiagnosticId.AlwaysUseVar.ToDiagnosticId(),
             Title,
             MessageFormat,
@@ -26,7 +26,18 @@ namespace CodeCracker.CSharp.Style
             description:Description,
             helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.AlwaysUseVar));
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+        internal static readonly DiagnosticDescriptor RulePrimitives = new DiagnosticDescriptor(
+            DiagnosticId.AlwaysUseVarOnPrimitives.ToDiagnosticId(),
+            Title,
+            MessageFormat,
+            Category,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: Description,
+            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.AlwaysUseVarOnPrimitives));
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(RuleNonPrimitives, RulePrimitives);
 
         public override void Initialize(AnalysisContext context) =>
             context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.LocalDeclarationStatement);
@@ -54,8 +65,33 @@ namespace CodeCracker.CSharp.Style
                 if (!conversion.IsIdentity) return;
             }
 
-            var diagnostic = Diagnostic.Create(Rule, variableDeclaration.Type.GetLocation());
+            var rule = IsPrimitvie(variableType) ? RulePrimitives : RuleNonPrimitives;
+            var diagnostic = Diagnostic.Create(rule, variableDeclaration.Type.GetLocation());
             context.ReportDiagnostic(diagnostic);
+        }
+
+        private static bool IsPrimitvie(ITypeSymbol typeSymbol)
+        {
+            switch (typeSymbol.SpecialType)
+            {
+                case SpecialType.System_Boolean:
+                case SpecialType.System_Byte:
+                case SpecialType.System_SByte:
+                case SpecialType.System_Int16:
+                case SpecialType.System_UInt16:
+                case SpecialType.System_Int32:
+                case SpecialType.System_UInt32:
+                case SpecialType.System_Int64:
+                case SpecialType.System_UInt64:
+                case SpecialType.System_IntPtr:
+                case SpecialType.System_UIntPtr:
+                case SpecialType.System_Char:
+                case SpecialType.System_Double:
+                case SpecialType.System_Single:
+                    return true;
+                default:
+                    return false;
+            }
         }
     }
 }

--- a/src/CSharp/CodeCracker/Style/AlwaysUseVarCodeFixProvider.cs
+++ b/src/CSharp/CodeCracker/Style/AlwaysUseVarCodeFixProvider.cs
@@ -17,7 +17,7 @@ namespace CodeCracker.CSharp.Style
     public class AlwaysUseVarCodeFixProvider : CodeFixProvider
     {
         public sealed override ImmutableArray<string> FixableDiagnosticIds =>
-            ImmutableArray.Create(DiagnosticId.AlwaysUseVar.ToDiagnosticId());
+            ImmutableArray.Create(DiagnosticId.AlwaysUseVar.ToDiagnosticId(), DiagnosticId.AlwaysUseVarOnPrimitives.ToDiagnosticId());
 
         public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
 

--- a/src/Common/CodeCracker.Common/DiagnosticId.cs
+++ b/src/Common/CodeCracker.Common/DiagnosticId.cs
@@ -79,5 +79,6 @@
         ConsoleWriteLine = 95,
         NameOf_External = 108,
         StringFormatArgs_ExtraArgs = 111,
+        AlwaysUseVarOnPrimitives = 105,
     }
 }

--- a/test/CSharp/CodeCracker.Test/Style/AlwaysUseVarTests.cs
+++ b/test/CSharp/CodeCracker.Test/Style/AlwaysUseVarTests.cs
@@ -7,6 +7,7 @@ namespace CodeCracker.Test.CSharp.Style
 {
     public class AlwaysUseVarTests : CodeFixVerifier<AlwaysUseVarAnalyzer, AlwaysUseVarCodeFixProvider>
     {
+
         [Fact]
         public async Task IgnoresConstantDeclarations()
         {
@@ -93,10 +94,12 @@ namespace CodeCracker.Test.CSharp.Style
 
         }
 
-        [Fact]
-        public async Task CreateDiagnosticsWhenAssigningValueWithSameDeclaringType()
+        [Theory]
+        [InlineData("int")]
+        [InlineData("System.Int32")]
+        public async Task CreateDiagnosticsWhenAssigningValueWithSameDeclaringTypePrimitive(string value)
         {
-            const string test = @"
+            var test = @"
     using System;
 
     namespace ConsoleApplication1
@@ -105,7 +108,36 @@ namespace CodeCracker.Test.CSharp.Style
         {
             public async Task Foo()
             {
-                int a = 10;
+                " + value + @" a = 10;
+            }
+        }
+    }";
+            var expected = new DiagnosticResult
+            {
+                Id = DiagnosticId.AlwaysUseVarOnPrimitives.ToDiagnosticId(),
+                Message = "Use 'var' instead of specifying the type name.",
+                Severity = DiagnosticSeverity.Warning,
+                Locations = new[] { new DiagnosticResultLocation("Test0.cs", 10, 17) }
+            };
+
+            await VerifyCSharpDiagnosticAsync(test, expected);
+        }
+
+        [Theory]
+        [InlineData("string a = \"10\";")]
+        [InlineData("DateTime date1 = new DateTime(2013, 6, 1, 12, 32, 30);")]
+        public async Task CreateDiagnosticsWhenAssigningValueWithSameDeclaringTypeNonPrimitive(string value)
+        {
+            var test = @"
+    using System;
+
+    namespace ConsoleApplication1
+    {
+        class TypeName
+        {
+            public async Task Foo()
+            {
+                " + value + @"
             }
         }
     }";

